### DISCUSSION TITLE
如何在保持杜比视界元数据的情况下合并下载下来的视频和音频文件

### DISCUSSION THREAD

**Discussion by SwimmingTiger** [the1812/Bilibili-Evolved#3500] [提问 / Q&A] [answered]
比如这个视频：https://www.bilibili.com/video/BV1eV411W7tt
下载下来的视频和音频是两个单独的文件，视频文件用电视播放显示杜比视界，音频文件播放显示杜比全景声。

![image](https://user-images.githubusercontent.com/4986069/179193655-2fc80463-c839-4173-8bbc-4eda67246a27.png)

用`ffmpeg`合并：
```
ffmpeg -i 375186693-1-30126.mp4 -i 375186693-1-30250.mp4 -c copy -map 0:v:0 -map 1:a:0 merged.mp4
```
丢失杜比视界元数据，只显示杜比全景声，并且视频偏色。

谁知道怎么合并才能保持杜比视界元数据？就是播放时同时出现杜比视界和杜比全景声。

**Comment by SwimmingTiger** [comment]
从网上找到一个答案：https://www.reddit.com/r/ffmpeg/comments/qe7oq1/dolby_vision_from_mkv_to_mp4_using_ffmpeg_and/

1. 安装`ffmpeg`：https://ffmpeg.org/download.html
2. 安装`mp4muxer`：https://github.com/DolbyLaboratories/dlb_mp4base/tree/master/bin
3. 运行如下命令：
   ```
   ffmpeg -i 待合并的视频.mp4 -c copy video.hevc
   ffmpeg -i 待合并的音频.mp4 -c copy audio.eac3
   ```
4. 把`audio.eac3`重命名为`audio.ec3`。

   为什么在上面的命令里不把文件名直接改成`.ec3`？
   因为`ffmpeg`通过文件扩展名决定默认参数，写`.eac3`可以免掉一大堆参数。
   如果写`.ec3`又不加音频格式参数，就会报未知格式错误。
   
   那为什么要重命名？
   因为`mp4muxer`只支持`.ec3`，输入`.eac3`会报未知扩展名错误。
5. 对于 PQ + Profile 5 的杜比视界文件，运行如下命令：
   ```
   mp4muxer --dv-profile 5 -i video.hevc -i audio.ec3 -o outputfile.mp4
   ```

   对于 HLG + Profile 8.4 的杜比视界文件，运行如下命令：
   ```
   mp4muxer --dv-profile 8 --dv-bl-compatible-id 4 -i video.hevc -i audio.ec3 -o outputfile.mp4
   ```

亲测合并后电视同时显示杜比视界和杜比全景声标识。

![IMG_20220715_182839](https://user-images.githubusercontent.com/4986069/179206104-f2f0fd6b-f3ac-4b43-9828-33d18f7741bc.jpg)

### 如何区分 Profile 5 和 Profile 8.4？

在执行`ffmpeg -i 待合并的视频.mp4 -c copy video.hevc`命令时，如果输出包含：

> encoder         : DOVI Coding

就是 Profile 5。

![image](https://user-images.githubusercontent.com/4986069/179204369-a23cc60b-247c-478d-8507-cde2ae302aa9.png)

如果是：

> encoder         : HEVC Coding

就是 Profile 8.4。

![image](https://user-images.githubusercontent.com/4986069/179204411-8833a20e-248e-48a6-97b2-d6dbedadefbb.png)

**Reply by SwimmingTiger** [reply]
LG CX电视自带的播放器不支持ass字幕的移动轨迹定义，弹幕ass文件播出来是这样的，气笑了。

![IMG_20220715_184605_edit_263600385741548](https://user-images.githubusercontent.com/4986069/179209020-aa990452-edc2-4485-9970-c3a673c399ae.jpg)

**Comment by lns103** [comment, marked as answer]
我使用这条命令，比起楼上的方法更简洁 ```ffmpeg -i video.m4s -i audio.m4s -strict unofficial -c copy output.mp4```

**Reply by lns103** [reply]
去年我在https://github.com/the1812/Bilibili-Evolved/discussions/2161 提及过，当时网络上还找不到相关内容，我了解到是side data问题之后，在网络上搜索到360°全景视频合并可以这样保留side data，然后在杜比视界上测试通过

**Reply by SwimmingTiger** [reply]
确实有效，输出结果可以激活电视的杜比视界和杜比全景声。

不过我发现 ffmpeg 和 mp4muxer 都改变了原视频的编码器类型参数。


原视频，电视认为编码器是DVH1：
![IMG_20220715_205947_edit_267012800967589](https://user-images.githubusercontent.com/4986069/179228100-4139bc5e-33a9-4904-aa17-3909fa13e167.jpg)


ffmpeg的输出，电视认为编码器是HEVC：
![IMG_20220715_204315_edit_266139748044285](https://user-images.githubusercontent.com/4986069/179226382-ec4f7fab-fe7e-47f6-af5a-76ce15c58ce3.jpg)

mp4muxer的输出，电视认为编码器是DVHE：
![IMG_20220715_204433_edit_266120251392205](https://user-images.githubusercontent.com/4986069/179226393-1924fde3-8d18-420a-a2ae-c5565c194f94.jpg)

画面好像是一模一样的，所以电视解码杜比视界时应该不依靠这个编码器类型。

**Reply by phoukn** [reply]
万分感谢啊

**Comment by 12138xcv-oss** [comment]
大佬，这个mp4muxer怎么安装？

**Reply by SwimmingTiger** [reply]
https://github.com/bbeny123/mp4muxer/releases

**Reply by SwimmingTiger** [reply]
用lns103的方法更好：
```
ffmpeg -i video.m4s -i audio.m4s -strict unofficial -c copy output.mp4
```
